src/tmpfiles/tmpfiles.c

```diff
@@ -2004,6 +2006,64 @@ static int glob_item_recursively(Item *i, fdaction_t action) {
         return r;
 }
 
+static int rm_if_wrong_type(mode_t mode, const char *path, bool follow) {
+        struct stat st;
+        int r;
+
+        assert(path);
+        assert((mode & ~S_IFMT) == 0);
+        log_debug("Checking type of %s", path);
+
+        if (follow)
+                r = stat(path, &st);
+        else
+                r = lstat(path, &st);
+        if (r < 0) {
+                if (errno != ENOENT) {
+                        log_error_errno(r, "stat(%s): %m", path);
+                }
+                return -errno;
+        }
+
+        if ((st.st_mode & S_IFMT) == mode)
+                return 0;
+
+        log_warning("wrong file type 0x%x; rm -rf \"%s\"", st.st_mode & S_IFMT, path);
+        r = rm_rf(path, REMOVE_ROOT|REMOVE_SUBVOLUME|REMOVE_PHYSICAL);
+        if (r < 0) {
+                if (errno != ENOENT) {
+                        log_error_errno(r, "rm_rf(%s): %m", path);
+                }
+                return -errno;
+        }
+
+        return 0;
+}
+
+static int rm_nondir_parents(const char *path) {
+        char *p, *e;
+        int r = -ENOENT;
+
+        assert(path);
+
+        // Walk up the path components until one is found with the expected type or there are no more.
+        p = strdupa(path);
+        // If the path doesn't exist, check the next path component.
+        while (r == -ENOENT || r == -ENOTDIR) {
+                e = strrchr(p, '/');
+                if (!e)
+                        return 0;
```

Comment:
in new code, please use path_extract_directory() to find the parent of a path, let's not handroll this anymore.

Reply:
I will take a look. This will probably look a bit different after the switch to openat.

Reply:
I didn't find a helper/macro for walking through the path components.